logging: Add option to block in thread context

nordic-krch opened the pull request:
Added CONIFG_LOG_BLOCK_IN_THREAD option to block until buffer for
log message is available. When log message is called in the thread
and there is no buffer available in the pool, thread will block with
configurable timeout (CONFIG_LOG_BLOCK_IN_THREAD_TIMEOUT_MS). If
buffer cannot be allocated by that time, message will be dropped.

Fixes #15196.

nordic-krch committed:
logging: Add option to block in thread context

Added CONIFG_LOG_BLOCK_IN_THREAD option to block until buffer for
log message is available. When log message is called in the thread
and there is no buffer available in the pool, thread will block with
configurable timeout (CONFIG_LOG_BLOCK_IN_THREAD_TIMEOUT_MS). If
buffer cannot be allocated by that time, message will be dropped.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch commented:
fyi @jimparis

dbkinder commented on doc/reference/logging/index.rst:
```diff
@@ -95,6 +95,10 @@ oldest one are discarded.
 
+:option:`CONFIG_LOG_BLOCK_IN_THREAD`: If enabled then when logger cannot
+allocate new message in thread context it will block for
+:option:`LOG_BLOCK_IN_THREAD_TIMEOUT_MS`.
```
```suggestion
:option:`CONFIG_LOG_BLOCK_IN_THREAD_TIMEOUT_MS` milliseconds.
```

dbkinder reviewed (changes requested):
fix typo in option name (causing a doc build warning), and clarify sentence a bit.

pizi-nordic commented on doc/reference/logging/index.rst:
```diff
@@ -95,6 +95,11 @@ oldest one are discarded.
 :option:`CONFIG_LOG_MODE_NO_OVERFLOW`: When logger cannot allocate new message
 it is discarded.
 
+:option:`CONFIG_LOG_BLOCK_IN_THREAD`: If enabled and new log message cannot
```
I would add some kind of "low water mark" below which thread will block. 
Such mechanism will leave some memory for contexts we cannot block.

nordic-krch replied:
It's quite difficult to do that. Currently, I used built-in feature of `mem_slab` which blocks when it's empty. What you suggested is more complex. I would prefer to get that in and consider "log water mark" in the future.

pizi-nordic replied:
Ok. Could you please fill the issue suggesting this enhancement?

pizi-nordic replied:
Ping

nordic-krch replied:
https://github.com/zephyrproject-rtos/zephyr/issues/19328

pizi-nordic commented on include/logging/log_msg.h:
```diff
@@ -22,6 +22,10 @@ extern "C" {
  */
 
+#ifndef CONFIG_LOG_BLOCK_IN_THREAD_TIMEOUT_MS
+#define CONFIG_LOG_BLOCK_IN_THREAD_TIMEOUT_MS 0
```
#define CONFIG_LOG_BLOCK_IN_THREAD_TIMEOUT_MS K_NO_WAIT

nordic-krch replied:
i wouldn't do that. It is only to satisfy compiler when feature is disabled. Changing it to `K_NO_WAIT` could suggest that it is used. I'll add comment.

pizi-nordic reviewed (changes requested):
Unconditional sleeping in  thread could lead to the similar issues as these reported in https://github.com/zephyrproject-rtos/zephyr/issues/16273

nordic-krch commented:
> Unconditional sleeping in thread could lead to the similar issues as these reported in #16273

are you referring to mem_slab allocation with timeout? I guess we can expect to get that bug fixed and when fixed `k_is_in_irq` will need to be replaced with `k_is_in_atomic`.

jhedberg commented:
Would it be possible to get the same feature for `log_strdup()` as well?

nordic-krch commented:
@jhedberg yes that would be possible.
Anyway, I'll put this PR on hold for now until #16273 is resolved since with this bug adding blocking in logger which can be called from any context is too risky.

dbkinder reviewed (approved):
doc changes LGTM, thanks.

nordic-krch commented:
rebased, I've also modified decision function. It's checking if in preemptive thread and if yes it checks if interrupts are locked. @pizi-nordic please take a look.

jhedberg commented:
> It's checking if in preemptive thread and if yes it checks if interrupts are locked. 

What's the problem with blocking in cooperative threads? Giving a non-zero timeout to kernel APIs in a cooperative thread should still let other threads continue their work. If there's something I'm missing and you really do need to limit this only to preemptible threads I think you should update the Kconfig help text to mention this.

nordic-krch commented:
@jhedberg changed to `!is_in_isr()`

jhedberg commented on subsys/logging/log_msg.c:
```diff
@@ -43,10 +49,34 @@ void log_msg_pool_init(void)
+}
+
+/* Check if context can be blocked and pend on available memory slab. Context
+ * can be blocked if in preemptive thread and interrupts are not locked.
```
Here's still a left-over reference to preemptive threads

nordic-krch replied:
no, it's not 😉 fixed.

jhedberg reviewed (changes requested):
> @jhedberg changed to !is_in_isr()

Thanks, that seems better. There's one comment that still needs updating though.

pizi-nordic reviewed (approved):
Please fill an issue about: https://github.com/zephyrproject-rtos/zephyr/pull/16194/files#r285938830

nordic-krch commented:
@nashif @carlescufi @jhedberg @andrewboie someone can merge?
